### PR TITLE
auto-enable -Werror based on compiler whitelist

### PR DESCRIPTION
**Pull request by tridge**
This fixes warnings then locks it in by enabling -Werror for specific white listed compilers in STM32 builds

### CODE REVIEW

**Comment by tridge** [comment]
another way of suppressing uavcan errors #15265 
we'll see which one passes CI :-)

**Review by peterbarker — APPROVED** [review]
I think I prefer @bugobliterator 's fix with the push's.  At least if we ever work out how to do things without the warning there are fewer places to change then.

**Comment by magicrub** [comment]
> ../../libraries/AP_Airspeed/AP_Airspeed_UAVCAN.cpp:18:32: error: unknown option after â€˜#pragma GCC diagnosticâ€™ kind [-Werror=pragmas]
 #pragma GCC diagnostic ignored "-Wcast-function-type"
                               ^
compilation terminated due to -Wfatal-errors.

**Comment by tridge** [comment]
this is not passing yet due to very old compiler in CI

**Comment by tridge** [comment]
I've tried switching travis to focal (ubuntu 20.04) for a newer compiler

**Comment by tridge** [comment]
related PR: #15114

**Comment by tridge** [comment]
I've reworked this to use a whitelisted set of compilers to enable with -Werror for STM32 boards. This means most devs and CI will end up with -Werror for most boards

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -622,6 +622,16 @@ def configure_env(self, cfg, env):
             cfg.srcnode.find_dir('libraries/AP_GyroFFT/CMSIS_5/include').abspath()
         ]
 
+        # whitelist of compilers which we should build with -Werror
+        gcc_whitelist = [
+            ('4','9','3'),
+            ('6','3','1')
```
Comment (peterbarker) [inline Tools/ardupilotwaf/boards.py:628]: Trailing comma here simplifies future diffs.

File: Tools/ardupilotwaf/boards.py [hunk 2]
```diff
@@ -622,6 +622,16 @@ def configure_env(self, cfg, env):
...
+        ]
+
+        if cfg.env.CC_VERSION in gcc_whitelist:
+            cfg.msg("Enabling -Werror", "yes")
```
[the remark this answers is not in the document]
Comment (tridge) [reply Tools/ardupilotwaf/boards.py:632]: the complication is the interaction with --Werror configure option. I've changed it to check for that

File: Tools/ardupilotwaf/boards.py (hunk 1 above)
Comment (tridge) [reply Tools/ardupilotwaf/boards.py]: done

**Pull request merged**